service/controller/resource/deployment/serviceendpoints.go

```diff
@@ -0,0 +1,57 @@
+package deployment
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-11-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/reconciliationcanceledcontext"
+
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+const (
+	storageServiceEndpoint = "Microsoft.Storage"
+)
+
+func (r *Resource) ensureServiceEndpoints(ctx context.Context, cr providerv1alpha1.AzureConfig) error {
+	subnetsClient, err := r.clientFactory.GetSubnetsClient(cr.Spec.Azure.CredentialSecret.Namespace, cr.Spec.Azure.CredentialSecret.Name)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	subnet, err := subnetsClient.Get(ctx, key.ResourceGroupName(cr), key.VnetName(cr), key.MasterSubnetName(cr), "")
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if subnet.ServiceEndpoints != nil {
+		for _, existing := range *subnet.ServiceEndpoints {
+			if *existing.Service == storageServiceEndpoint {
+				return nil
+			}
+		}
+	} else {
+		subnet.ServiceEndpoints = &[]network.ServiceEndpointPropertiesFormat{}
+	}
+
+	r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("service '%s' for subnet %s was not enabled. Adding it.", storageServiceEndpoint, *subnet.Name))
+
+	*subnet.ServiceEndpoints = append(*subnet.ServiceEndpoints, network.ServiceEndpointPropertiesFormat{
+		Service: to.StringPtr(storageServiceEndpoint),
+	})
+
+	_, err = subnetsClient.CreateOrUpdate(ctx, key.ResourceGroupName(cr), key.VnetName(cr), key.MasterSubnetName(cr), subnet)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("service '%s' was added to subnet %s", storageServiceEndpoint, *subnet.Name))
+
+	reconciliationcanceledcontext.SetCanceled(ctx)
```

Comment:
why do we have to cancel the reconciliation? Now that we ensured the endpoints, can't we just continue?

Reply:
when the endpoint gets created, the virtual network is in an "updating" state for a while.
If we don't cancel the reconciliation, the subsequent deployment applications (it's the very next instruction so it's a quick thing) fails because azure is a liquid stinky shit.
So we have no alternative but wait one reconciliation loop if we change the service endpoints.